[Core] Add UnusedStepsSummaryPrinter

timtebeek opened the pull request:
<!-- These sections are meant as guidance for you. If something doesn't fit, you can just skip it. -->

## Summary

Adds a plugin to find and report on unused steps.

## Details

As discussed in https://github.com/cucumber/cucumber-jvm/pull/1634 it uses the new StepDefinedEvent to locate all registered steps, and prints a summary of unused steps to the argument file.

## Motivation and Context

Allows to easily find unused steps, which might be indicative of removed functionality or missed test coverage.

## How Has This Been Tested?

Not yet; as @mpkorstanje indicated some hesitation on whether this would warrant adoption. This PR is to record that discussion, and will be updated depending on the outcome. Same goes for the current Java8 baseline, so it might need to target a future version rather than the next.

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

mpkorstanje commented:
Could you show me an example of the output?

timtebeek commented:
The current output is:
```
2 Unused steps:
SampleSteps.java:17: this step is unused
SampleSteps.maar(): another unused step
```
For a sample project here: https://github.com/timtebeek/cucumber-jvm-issues-1633/tree/step-defined-event/src/test/java/com/github/timtebeek

mpkorstanje commented:
Looks good. I was initially thinking we should use the same format as the summary printer, but we may as well add it to the summary printer. That way there is no need for an additional `--dry-run` or separate configuration (for most use cases).

So let's add it to the default summary printer! 

For example:

```
Undefined scenarios:
io/cucumber/skeleton/belly.feature:3 # a few cukes
io/cucumber/skeleton/belly2.feature:3 # a few cukes

Unused steps:
SampleSteps.java:17 # this step is unused
SampleSteps.maar() # another unused step

2 Scenarios (2 undefined)
6 Steps (4 undefined, 2 passed)
2 Unused steps 
0m0.138s
```

* The separator is a hash rather then then a colon. 
* If there are no unused steps those sections need not be printed. 
* The locations should be printed in yellow unless is enabled.

![image](https://user-images.githubusercontent.com/6946919/58829824-197f0480-8649-11e9-8ff0-56acb7f332fc.png)

mpkorstanje commented:
Okay. That was a terrible idea. Its nice when running the full suite but a nuisance when running a selection.

Let's not add it, but let's use the same format.

timtebeek commented:
Running a selection is indeed why I had implemented this as a separate plugin so far; thanks for the suggestions; I've updated the plugin accordingly. Not really sure how to cover this with tests, and/or if we're still on track for an eventual merge..

mpkorstanje commented:
We are good. I think a simple unit test where events go in and and the right output string is produced should be sufficient.

timtebeek commented:
Perfect; added `UnusedStepsSummaryPrinterTest` in https://github.com/cucumber/cucumber-jvm/pull/1648/commits/679dd3f5992ee3b3d06ac0a6d579dbe71f2caa63 , hope that matches what you were looking for. Anything else before we can consider a merge?

mpkorstanje commented on core/src/main/java/cucumber/runtime/formatter/UnusedStepsSummaryPrinter.java:
```diff
@@ -0,0 +1,76 @@
+
+	final Map<String, String> unusedSteps = new TreeMap<>();
+	final NiceAppendable out;
+	Formats formats = new MonochromeFormats();
```
`private` is missing I guess.

mpkorstanje replied:
```suggestion
	private Formats formats = new MonochromeFormats();
```

mpkorstanje commented on core/src/main/java/cucumber/runtime/formatter/UnusedStepsSummaryPrinter.java:
```diff
@@ -0,0 +1,76 @@
+			for (Entry<String, String> entry : unusedSteps.entrySet()) {
+				String location = entry.getKey();
+				String pattern = entry.getValue();
+				out.println(format.text(location + " # " + pattern));
```
```suggestion
				out.println(format.text(location) + " # " + pattern);
```

Only the location is in yellow.

mpkorstanje commented on core/src/main/java/cucumber/runtime/formatter/UnusedStepsSummaryPrinter.java:
```diff
@@ -0,0 +1,76 @@
+		}
+	};
+
+	final Map<String, String> unusedSteps = new TreeMap<>();
```
```suggestion
	private final Map<String, String> unusedSteps = new TreeMap<>();
```

mpkorstanje commented on core/src/main/java/cucumber/runtime/formatter/UnusedStepsSummaryPrinter.java:
```diff
@@ -0,0 +1,76 @@
+	};
+
+	final Map<String, String> unusedSteps = new TreeMap<>();
+	final NiceAppendable out;
```
```suggestion
	private final NiceAppendable out;
```

timtebeek commented:
Thanks for directly applying your own suggestions.. as a note: I'd used package private modifier as the inner class event handlers otherwise need to use syntactic accessors to access the private fields (or so my IDE tells me), but the code should work fine as it is now, and probably clearer.

Shall I merge? Or feel free to merge of that's faster.. :) great to see this added, saves me copy and pasting it where I need it.

mpkorstanje commented:
It's a new github feature. Really nice. 

I've merged it. I've been using a format that makes it look reasonably nice. But I still need to document it so other people can use it too.

timtebeek commented:
Perfect, thanks for the review/guidance/merge! :)

mpkorstanje commented:
You're welcome. Anytime.
